### ISSUE TITLE
Totally-partial return type treated as an empty interface

### ISSUE THREAD

**Issue by ds300** [Microsoft/TypeScript#16254]
**TypeScript Version:**  2.3.4

**Code**

```ts
// A *self-contained* demonstration of the problem follows...

// Let's say we have a type with all optional properties
interface A {
  a?: boolean
}

// we can't assign types of the wrong shape to A
const a: A = {b: true} // correctly identified as error

// But what if we want to deal with functions that return that type
function evalThunk(thunk: () => A): A {
  return thunk()
}

// here types of the wrong shape are assignable to A
evalThunk(() => ({b: true})) // no error

```

**Expected behavior:**

Expected the return type of functions to be checked with the same semantics as parameter assignment

**Actual behavior:**

Seems like the semantics differ. Probably it somehow treats the return type like being equivalent to the empty interface, since all properties are optional.

Note that this also happens with `Partial<T>` types. And also that if you add a non-optional property to the `A` interface, the error is correctly raised.

Apologies if this has been reported before.

**Comment by jwbay** [comment]
Might be fixed by #16047? Not sure

**Comment by ahejlsberg** [comment]
This is indeed fixed by #16047.

**Comment by ds300** [comment]
Thanks, I'll close this then
